[BUG] `negative_data_rejection` false positive for URL parameters

### Checklist

- [x] I checked the [FAQ section](https://schemathesis.readthedocs.io/en/stable/faq.html#frequently-asked-questions) of the documentation
- [x] I looked for similar issues in the [issue tracker](https://github.com/schemathesis/schemathesis/issues)
- [x] I am using the latest version of Schemathesis

### Describe the bug

I really like that you added negative data rejection, it's a really valuable addition to this library.

Unfortunately, it fails for URL parameters when it shouldn't.

E.g. if you have a Path like `/projects/?slug={slug}` or `/project/{slug}`, with `slug` being a string, it will generate negative examples like `null` or `false`, which is perfectly sensible for JSON payloads, but not for URLs, since everything there is a string. And for a slug of a project, `"false"` could be an acceptable project slug for a project called `"False"`. 
So an API accepting `false` in this case is valid, and should not count as failed negative data rejection.

Another, related but slightly different bug is with this API spec: 

```
  /namespaces:
    get:
      summary: Get all namespaces
      parameters:
        - in: query
          description: Result's page number starting from 1
          name: page
          required: false
          schema:
            type: integer
            minimum: 1
            default: 1
        - in: query
          description: The number of results per page
          name: per_page
          required: false
          schema:
            type: integer
            minimum: 1
            maximum: 100
            default: 20
      responses:
        "200":
          description: List of namespaces
```
which fails with
```
E       schemathesis.exceptions.CheckFailed:
E
E       1. Accepted negative data
E
E           Negative data was not rejected as expected by the API
E
E       [200] OK:
E
E           `[{"id":"01J1SA7WBQA366YZHNA33Q8P9G","name":"Admin Doe","slug":"admin.doe","created_by":"admin","namespace_kind":"user"},{"id":"01J1SAQ4BG7915N27WKVC5N1AJ","name":"\udbff\udfb8\u00b8Q\u00cb\u00ab\u00eb","slug":"pz","created_by":"admin","namespace_kind":"group"}]`
E
E       Reproduce with:
E
E           curl -X GET -H 'host: mockserver:1234' -H 'authorization: Bearer {"is_admin": true, "id": "admin", "name": "Admin Doe", "first_name": "Admin", "last_name": "Doe", "email": "admin.doe@gmail.com", "full_name": "Admin Doe"}' 'http://localhost/api/data/namespaces?=null&page=1'
E
E       Falsifying example: async_run(
E           sanic_client=<sanic_testing.testing.SanicASGITestClient object at 0x73c6ad790260>,
E           admin_headers={'Authorization': 'Bearer {"is_admin": true, "id": "admin", "name": "Admin Doe", "first_name": "Admin", "last_name": "Doe", "email": "admin.doe@gmail.com", "full_name": "Admin Doe"}'},
E           case=Case(query={'': 'null', 'page': 1}),
E       )
```
which is a bit weird test data and I think most servers will just ignore an empty key in the URL, so not really a failure.

The actual tests that we run can be found at https://github.com/SwissDataScienceCenter/renku-data-services/blob/11029e0fcc0f8e0298448eacd67d1314fff37ec2/test/bases/renku_data_services/data_api/test_schemathesis.py

### To Reproduce

🚨 **Mandatory** 🚨: Steps to reproduce the behavior:

Apispec like
```
---
openapi: 3.0.2
paths:
  "/project/{slug}":
    get:
      parameters:
        - in: path
          name: slug
          required: true
          schema:
            type: string
      responses:
        "200":
          content:
            text/plain:
              schema:
                type: string
        default:
          content:
            text/string:
              schema:
                type: string
```

with an API that accepts string as slug.

This will fail with something like:

```
schemathesis.exceptions.CheckFailed: 
  E       
  E       1. Accepted negative data
  E       
  E           Negative data was not rejected as expected by the API
  E       
  E       [200] OK:
  E       
  E           `...`
  E       
  E       Reproduce with: 
  E       
  
  E           curl -X DELETE -H 'host: mockserver:1234' -H 'authorization: ***"is_admin": true, "id": "admin", "name": "Admin Doe", "first_name": "Admin", "last_name": "Doe", "email": "admin.doe@gmail.com", "full_name": "Admin Doe"}' 'http://localhost/project/false'
  E       
  E       Falsifying example: async_run(
  E           sanic_client=<sanic_testing.testing.SanicASGITestClient object at 0x7f6fa5123f80>,
  E           admin_headers={'Authorization': '***"is_admin": true, "id": "admin", "name": "Admin Doe", "first_name": "Admin", "last_name": "Doe", "email": "admin.doe@gmail.com", "full_name": "Admin Doe"}'},
   E           case=Case(query={'slug': 'false'}),
  E       )
```

### Expected behavior

URL parameter should have a reduce set of negative examples for strings, like empty string, but not things like `false`, `true`, `null`.

### Environment

```
- OS: Linux
- Python version: 3.12
- Schemathesis version: 3.31.0
- Spec version: Open API 3.0.2
```

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -8,6 +8,10 @@ Changelog
 
 - Generating negative test cases for path and query parameters. :issue:`2312`
 
+**Changed**
+
+- Do not consider ignoring additional parameters as a failure in ``negative_data_rejection``.
+
 .. _v3.31.0:
 
 :version:`3.31.0 <v3.30.4...v3.31.0>` - 2024-06-30

--- a/src/schemathesis/models.py
+++ b/src/schemathesis/models.py
@@ -119,6 +119,19 @@ def prepare_request_data(kwargs: dict[str, Any]) -> PreparedRequestData:
     )
 
 
+@dataclass
+class GenerationMetadata:
+    """Stores various information about how data is generated."""
+
+    query: DataGenerationMethod | None
+    path_parameters: DataGenerationMethod | None
+    headers: DataGenerationMethod | None
+    cookies: DataGenerationMethod | None
+    body: DataGenerationMethod | None
+
+    __slots__ = ("query", "path_parameters", "headers", "cookies", "body")
+
+
 @dataclass(repr=False)
 class Case:
     """A single test case parameters."""
@@ -138,6 +151,8 @@ class Case:
     # The media type for cases with a payload. For example, "application/json"
     media_type: str | None = None
     source: CaseSource | None = None
+
+    meta: GenerationMetadata | None = None
 
     # The way the case was generated (None for manually crafted ones)
     data_generation_method: DataGenerationMethod | None = None

--- a/src/schemathesis/specs/openapi/_hypothesis.py
+++ b/src/schemathesis/specs/openapi/_hypothesis.py
@@ -25,7 +25,7 @@ from ...generation import DataGenerationMethod, GenerationConfig
 from ...hooks import HookContext, HookDispatcher, apply_to_all_dispatchers
 from ...internal.copy import fast_deepcopy
 from ...internal.validation import is_illegal_surrogate
-from ...models import APIOperation, Case, cant_serialize
+from ...models import APIOperation, Case, GenerationMetadata, cant_serialize
 from ...serializers import Binary
 from ...transports.content_types import parse_content_type
 from ...transports.headers import has_invalid_characters, is_latin_1_encodable
@@ -36,7 +36,7 @@ from .formats import STRING_FORMATS
 from .media_types import MEDIA_TYPES
 from .negative import negative_schema
 from .negative.utils import can_negate
-from .parameters import OpenAPIBody, parameters_to_json_schema
+from .parameters import OpenAPIBody, OpenAPIParameter, parameters_to_json_schema
 from .utils import is_header_location
 
 HEADER_FORMAT = "_header_value"
@@ -207,6 +207,13 @@ def get_case_strategy(
         query=query_.value,
         body=body_.value,
         data_generation_method=generator,
+        meta=GenerationMetadata(
+            query=query_.generator,
+            path_parameters=path_parameters_.generator,
+            headers=headers_.generator,
+            cookies=cookies_.generator,
+            body=body_.generator,
+        ),
     )
     auth_context = auths.AuthContext(
         operation=operation,
@@ -347,6 +354,22 @@ def can_negate_headers(operation: APIOperation, location: str) -> bool:
     return any(header != {"type": "string"} for header in headers.values())
 
 
+def get_schema_for_location(
+    operation: APIOperation, location: str, parameters: Iterable[OpenAPIParameter]
+) -> dict[str, Any]:
+    schema = parameters_to_json_schema(operation, parameters)
+    if location == "path":
+        if not operation.schema.validate_schema:
+            # If schema validation is disabled, we try to generate data even if the parameter definition
+            # contains errors.
+            # In this case, we know that the `required` keyword should always be `True`.
+            schema["required"] = list(schema["properties"])
+        for prop in schema.get("properties", {}).values():
+            if prop.get("type") == "string":
+                prop.setdefault("minLength", 1)
+    return operation.schema.prepare_schema(schema)
+
+
 def get_parameters_strategy(
     operation: APIOperation,
     strategy_factory: StrategyFactory,
@@ -361,17 +384,7 @@ def get_parameters_strategy(
         nested_cache_key = (strategy_factory, location, tuple(sorted(exclude)))
         if operation in _PARAMETER_STRATEGIES_CACHE and nested_cache_key in _PARAMETER_STRATEGIES_CACHE[operation]:
             return _PARAMETER_STRATEGIES_CACHE[operation][nested_cache_key]
-        schema = parameters_to_json_schema(operation, parameters)
-        if location == "path":
-            if not operation.schema.validate_schema:
-                # If schema validation is disabled, we try to generate data even if the parameter definition
-                # contains errors.
-                # In this case, we know that the `required` keyword should always be `True`.
-                schema["required"] = list(schema["properties"])
-            for prop in schema.get("properties", {}).values():
-                if prop.get("type") == "string":
-                    prop.setdefault("minLength", 1)
-        schema = operation.schema.prepare_schema(schema)
+        schema = get_schema_for_location(operation, location, parameters)
         for name in exclude:
             # Values from `exclude` are not necessarily valid for the schema - they come from user-defined examples
             # that may be invalid

--- a/src/schemathesis/specs/openapi/checks.py
+++ b/src/schemathesis/specs/openapi/checks.py
@@ -145,13 +145,46 @@ def negative_data_rejection(response: GenericResponse, case: Case) -> bool | Non
 
     if not isinstance(case.operation.schema, BaseOpenAPISchema):
         return True
-    if case.data_generation_method and case.data_generation_method.is_negative and 200 <= response.status_code < 300:
+    if (
+        case.data_generation_method
+        and case.data_generation_method.is_negative
+        and 200 <= response.status_code < 300
+        and not has_only_additional_properties_in_non_body_parameters(case)
+    ):
         exc_class = get_negative_rejection_error(case.operation.verbose_name, response.status_code)
         raise exc_class(
             failures.AcceptedNegativeData.title,
             context=failures.AcceptedNegativeData(message="Negative data was not rejected as expected by the API"),
         )
     return None
+
+
+def has_only_additional_properties_in_non_body_parameters(case: Case) -> bool:
+    # Check if the case contains only additional properties in query, headers, or cookies.
+    # This function is used to determine if negation is solely in the form of extra properties,
+    # which are often ignored for backward-compatibility by the tested apps
+    from ._hypothesis import get_schema_for_location
+
+    meta = case.meta
+    if meta is None:
+        # Ignore manually created cases
+        return False
+    if (meta.body and meta.body.is_negative) or (meta.path_parameters and meta.path_parameters.is_negative):
+        # Body or path negations always imply other negations
+        return False
+    validator_cls = case.operation.schema.validator_cls  # type: ignore[attr-defined]
+    for container in ("query", "headers", "cookies"):
+        meta_for_location = getattr(meta, container)
+        value = getattr(case, container)
+        if value is not None and meta_for_location is not None and meta_for_location.is_negative:
+            parameters = getattr(case.operation, container)
+            value_without_additional_properties = {k: v for k, v in value.items() if k in parameters}
+            schema = get_schema_for_location(case.operation, container, parameters)
+            if not validator_cls(schema).is_valid(value_without_additional_properties):
+                # Other types of negation found
+                return False
+    # Only additional properties are added
+    return True
 
 
 def use_after_free(response: GenericResponse, original: Case) -> bool | None:

--- a/src/schemathesis/specs/openapi/schemas.py
+++ b/src/schemathesis/specs/openapi/schemas.py
@@ -19,6 +19,7 @@ from typing import (
     Mapping,
     NoReturn,
     Sequence,
+    Type,
     TypeVar,
     cast,
 )
@@ -615,6 +616,12 @@ class BaseOpenAPISchema(BaseSchema):
     def get_tags(self, operation: APIOperation) -> list[str] | None:
         return operation.definition.raw.get("tags")
 
+    @property
+    def validator_cls(self) -> Type[jsonschema.Validator]:
+        if self.spec_version.startswith("3.1") and experimental.OPEN_API_3_1.is_enabled:
+            return jsonschema.Draft202012Validator
+        return jsonschema.Draft4Validator
+
     def validate_response(self, operation: APIOperation, response: GenericResponse) -> bool | None:
         responses = {str(key): value for key, value in operation.definition.raw.get("responses", {}).items()}
         status_code = str(response.status_code)
@@ -658,13 +665,9 @@ class BaseOpenAPISchema(BaseSchema):
         resolver = ConvertingResolver(
             self.location or "", self.raw_schema, nullable_name=self.nullable_name, is_response_schema=True
         )
-        if self.spec_version.startswith("3.1") and experimental.OPEN_API_3_1.is_enabled:
-            cls = jsonschema.Draft202012Validator
-        else:
-            cls = jsonschema.Draft4Validator
         with in_scopes(resolver, scopes):
             try:
-                jsonschema.validate(data, schema, cls=cls, resolver=resolver)
+                jsonschema.validate(data, schema, cls=self.validator_cls, resolver=resolver)
             except jsonschema.ValidationError as exc:
                 exc_class = get_schema_validation_error(operation.verbose_name, exc)
                 ctx = failures.ValidationErrorContext.from_exception(exc, output_config=operation.schema.output_config)

--- a/test/specs/openapi/test_checks.py
+++ b/test/specs/openapi/test_checks.py
@@ -1,6 +1,14 @@
 import pytest
 
-from schemathesis.specs.openapi.checks import ResourcePath, _is_prefix_operation
+import schemathesis
+from schemathesis.generation import DataGenerationMethod
+from schemathesis.models import Case, GenerationMetadata
+from schemathesis.specs.openapi.checks import (
+    ResourcePath,
+    _is_prefix_operation,
+    has_only_additional_properties_in_non_body_parameters,
+    negative_data_rejection,
+)
 
 
 @pytest.mark.parametrize(
@@ -64,3 +72,92 @@ from schemathesis.specs.openapi.checks import ResourcePath, _is_prefix_operation
 )
 def test_is_prefix_operation(lhs, lhs_vars, rhs, rhs_vars, expected):
     assert _is_prefix_operation(ResourcePath(lhs, lhs_vars), ResourcePath(rhs, rhs_vars)) == expected
+
+
+def build_metadata(path_parameters=None, query=None, headers=None, cookies=None, body=None):
+    return GenerationMetadata(path_parameters=path_parameters, query=query, headers=headers, cookies=cookies, body=body)
+
+
+@pytest.fixture
+def sample_schema(empty_open_api_3_schema):
+    empty_open_api_3_schema["paths"] = {
+        "/test": {
+            "post": {
+                "parameters": [
+                    {
+                        "in": "query",
+                        "name": "key",
+                        "schema": {"type": "integer", "minimum": 5},
+                    },
+                    {
+                        "in": "headers",
+                        "name": "X-Key",
+                        "schema": {"type": "integer", "minimum": 5},
+                    },
+                ]
+            }
+        }
+    }
+    return empty_open_api_3_schema
+
+
+@pytest.mark.parametrize(
+    "kwargs, expected",
+    (
+        ({}, False),
+        (
+            {"meta": build_metadata(body=DataGenerationMethod.negative)},
+            False,
+        ),
+        (
+            {
+                "query": {"key": 1},
+                "meta": build_metadata(query=DataGenerationMethod.negative),
+            },
+            False,
+        ),
+        (
+            {
+                "query": {"key": 1},
+                "headers": {"X-Key": 42},
+                "meta": build_metadata(query=DataGenerationMethod.negative),
+            },
+            False,
+        ),
+        (
+            {
+                "query": {"key": 5, "unknown": 3},
+                "meta": build_metadata(query=DataGenerationMethod.negative),
+            },
+            True,
+        ),
+        (
+            {
+                "query": {"key": 5, "unknown": 3},
+                "headers": {"X-Key": 42},
+                "meta": build_metadata(query=DataGenerationMethod.negative),
+            },
+            True,
+        ),
+    ),
+)
+def test_has_only_additional_properties_in_non_body_parameters(sample_schema, kwargs, expected):
+    schema = schemathesis.from_dict(sample_schema)
+    operation = schema["/test"]["POST"]
+    case = Case(operation=operation, generation_time=0.0, **kwargs)
+    assert has_only_additional_properties_in_non_body_parameters(case) is expected
+
+
+def test_negative_data_rejection_on_additional_properties(response_factory, sample_schema):
+    # See GH-2312
+    response = response_factory.requests()
+    schema = schemathesis.from_dict(sample_schema)
+    operation = schema["/test"]["POST"]
+    case = Case(
+        operation=operation,
+        generation_time=0.0,
+        meta=build_metadata(query=DataGenerationMethod.negative),
+        data_generation_method=DataGenerationMethod.negative,
+        query={"key": 5, "unknown": 3},
+    )
+    assert negative_data_rejection(response, case) is None

--- a/test/test_direct_access.py
+++ b/test/test_direct_access.py
@@ -40,7 +40,7 @@ def test_as_strategy(swagger_20):
     strategy = operation.as_strategy()
     assert isinstance(strategy, st.SearchStrategy)
     assert strategy.example() == Case(
-        operation, generation_time=ANY, data_generation_method=DataGenerationMethod.positive
+        operation, generation_time=ANY, data_generation_method=DataGenerationMethod.positive, meta=ANY
     )
 
 

--- a/test/test_hypothesis.py
+++ b/test/test_hypothesis.py
@@ -75,6 +75,7 @@ def test_get_examples(location, swagger_20):
         generation_time=ANY,
         data_generation_method=DataGenerationMethod.positive,
         media_type=media_type,
+        meta=ANY,
         **{container: expected},
     )
 

